numba/core/interpreter.py

```diff
@@ -79,6 +79,488 @@ def get_assignment_source(self, destname):
+            raise UnsupportedError(errmsg)
+        arg_var = setitem_stmt.value.args[1]
+        # Append the (key, value) pair.
+        kws.append((key_val, arg_var))
+        # Remove the __setitem__ getattr and call
+        new_body[search_start] = None
+        new_body[search_start + 1] = None
+        search_start += 2
+    return kws
+
+
+def _call_function_ex_replace_args_small(tuple_expr, new_body, buildtuple_idx):
+    """
+    Extracts the args passed as vararg
+    for CALL_FUNCTION_EX. This pass is taken when
+    n_args <= 30 and the bytecode looks like:
+
+        # Start for each argument
+        LOAD_FAST  # Load each argument.
+        # End for each argument
+        ...
+        BUILD_TUPLE # Create a tuple of the arguments
+
+    In the IR generated, the vararg refer
+    to a single build_tuple that contains all of the
+    args. In addition to returning the args, this
+    function updates new_body to remove all usage
+    of the tuple.
+    """
+    # Delete the build tuple
+    new_body[buildtuple_idx] = None
+    # Return the args.
+    return tuple_expr.items
+
+
+def _call_function_ex_replace_args_large(
+    vararg_stmt, body, new_body, search_end
+):
+    """
+    Extracts the args passed as vararg
+    for CALL_FUNCTION_EX. This pass is taken when
+    n_args > 30 and the bytecode looks like:
+
+        BUILD_TUPLE # Create a list to append to
+        # Start for each argument
+        LOAD_FAST  # Load each argument.
+        LIST_APPEND # Add the argument to the list
+        # End for each argument
+        ...
+        LIST_TO_TUPLE # Convert the args to a tuple.
+
+    In the IR generated, the tuple is created by concatenating
+    together several 1 elem tuples to an initial empty tuple.
+    We traverse backwards in the IR, collecting args, until we
+    find the original empty tuple. For example, the IR might
+    look like:
+
+        $orig_tuple = build_tuple(items=[])
+        $first_var = build_tuple(items=[Var(arg0, test.py:6)])
+        $next_tuple = $orig_tuple + $first_var
+        ...
+        $final_var = build_tuple(items=[Var(argn, test.py:6)])
+        $final_tuple = $prev_tuple + $final_var
+        $varargs_var = $final_tuple
+
+    It unclear if the extra assignment at the end is always present.
+    In addition to collecting and returning the original args, we also
+    delete any IR statments that uses any of the tuples from new_body.
+    """
+    errmsg = "CALL_FUNCTION_EX with **kwargs not supported"
+    # We traverse to the front of the block to look for the original
+    # tuple.
+    search_start = 0
+    total_args = []
+    if (
+        isinstance(vararg_stmt, ir.Assign)
+        and isinstance(vararg_stmt.value, ir.Var)
+    ):
+        target_name = vararg_stmt.value.name
+        # If there is an initial assignment, delete it
+        new_body[search_end] = None
+        search_end -= 1
+    else:
+        target_name = vararg_stmt.target.name
```

Comment:
Why is this needed and what would trigger this line?

Reply:
This may never be needed, but I wasn't sure the code will always produce an assignment (although I can't construct a case where it hasn't happened).

Basically the existing IR looks like:

```python
...
$final_var = build_tuple(items=[Var(argn, test.py:6)])
$final_tuple = $prev_tuple + $final_var
$varargs_var = $final_tuple
```
but it would be completely valid if instead the IR generated was
```python
...
$final_var = build_tuple(items=[Var(argn, test.py:6)])
$varargs_var = $prev_tuple + $final_var
```
so I was trying to protect against this ever changing.

Reply:
If you are confident this extra assignment can't occur I'm happy to remove it.

Reply:
The extra assignment is coming from https://github.com/numba/numba/blob/59fa2e335be68148b3bd72a29de3ff011430038d/numba/core/interpreter.py#L259-L260. 

Let's make this branch raise an error to indicate it is unreachable and add a comment using your description in https://github.com/numba/numba/pull/7866#discussion_r817233996.